libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -384,3 +558,160 @@ float AC_Autorotation::calc_speed_forward(void)
+
+    //Limits the maximum change in pitch attitude based on acceleration
+    if (_accel_target > _accel_out_last + _accel_max) {
+        _accel_target = _accel_out_last + _accel_max;
+    } else if (_accel_target < _accel_out_last - _accel_max) {
+        _accel_target = _accel_out_last - _accel_max;
+    }
+
+    //Limiting acceleration based on velocity gained during the previous time step
+    if (fabsf(_delta_speed_fwd) > _accel_max * _dt) {
+        _flag_limit_accel = true;
+    } else {
+        _flag_limit_accel = false;
+    }
+
+    if ((_flag_limit_accel && fabsf(_accel_target) < fabsf(_accel_out_last)) || !_flag_limit_accel) {
+        _accel_out = _accel_target;
+    } else {
+        _accel_out = _accel_out_last;
+    }
+    _accel_out_last = _accel_out;
+
+
+
+    //estimate flare effectiveness
+
+    if (_speed_forward <= (0.6*_flare_entry_speed) && fabsf(_avg_sink_deriv)<=0.005f && _avg_acc_z>= -(1.1*9.80665f) ) {
+        if (!_flare_complete) {
+            _flare_complete = true;
+            gcs().send_text(MAV_SEVERITY_INFO, "Flare_complete");
+        }
+    }
+
+    if (!_flare_complete) {
+        _pitch_target = atanf(-_accel_out/(GRAVITY_MSS * 100.0f))*(18000.0f/M_PI);
+        _pitch_target = constrain_float(_pitch_target, 0.0f, AP_ALPHA_TPP*100.0f);
+    } else {
+        _pitch_target *= 0.95f;
+    }
+
+
+}
+
+void AC_Autorotation::touchdown_controller()
+{
+    float _current_sink_rate = _inav.get_velocity_z_up_cms();
+    if (_radar_alt>=_ground_clearance) {
+        _desired_sink_rate = linear_interpolate(0.0f, _entry_sink_rate, _radar_alt, _ground_clearance, _entry_alt);
+    } else {
+        _desired_sink_rate = 0.0f;
+    }
+    // update forward speed for logging
+    _speed_forward = calc_speed_forward(); //(cm/s)
+
+    _collective_out =  constrain_value((_p_coll_tch.get_p(_desired_sink_rate - _current_sink_rate))*0.01f + _ff_term_hs, 0.0f, 1.0f);
+    col_trim_lpf.set_cutoff_frequency(_col_cutoff_freq);
+    _ff_term_hs = col_trim_lpf.apply(_collective_out, _dt);
+    set_collective(HS_CONTROLLER_COLLECTIVE_CUTOFF_FREQ);
+    _pitch_target *= 0.95f;
+}
+
+void AC_Autorotation::get_entry_speed()
+{
+    _flare_entry_speed = calc_speed_forward();
+}
+
+void AC_Autorotation::time_to_ground()
+{
+    if (_inav.get_velocity_z_up_cms() < 0.0f ) {
+        _time_to_ground = -(_radar_alt/_inav.get_velocity_z_up_cms());
+    } else {
+        _time_to_ground = _t_tch +1.0f;
+    }
+}
+
+void AC_Autorotation::init_est_radar_alt()
+{
+    // set descent rate filter cutoff frequency
+    descent_rate_lpf.set_cutoff_frequency(40.0f);
+
+    // Reset feed descent rate filter
+    descent_rate_lpf.reset(_inav.get_velocity_z_up_cms());
+
+    _radar_alt_calc = _radar_alt;
+    _radar_alt_prev = _radar_alt;
+    _est_alt = _radar_alt;
+
+}
+
+void AC_Autorotation::update_est_radar_alt()
+{
+    if (_using_rfnd) {
+        // continue calculating radar altitude based on the most recent update and descent rate
```

Comment:
I'm not convinced we need this.  See my comment near the top (heli.cpp) regarding the rangefinder height that we are passing into autorotation.  we are already getting an result from an interpolation based on inertial nav and then we ignore that result if the range finder is good.  I think we have an opportunity to simplify this section of the code and delete most of the code in this function.